### DISCUSSION TITLE
Thoughts about Avalonia

### DISCUSSION THREAD

**Discussion by FirehawkV21** [AvaloniaUI/Avalonia#13269] [General]
I've recently shipped a tool named RPG Maker Cook Tool Deluxe (paid, https://studioace.wordpress.com/rmcooktooldx/). After shipping this tool, I'd like to share my thoughts and feedback regarding Avalonia:

# Things I liked

- **It's very fast:** Not counting my code in this, I am pretty impressed with how fast it goes. Seeing the final app running on an AMD Ryzen 5 5600G (Windows 11, used .NET 8) at super speed is rather surprising. One issue I had is that if I run a Parallel loop can slow down the app (at least when debugging. Publishing the app doesn't show this issue).
- **Native AOT does wonders:** Not necessarily Avalonia's doing (I have seen apps being far speedier when compiled to native), but it makes the already fast app even better. I did see trimming errors, but the app works.
- **UI is perfect:** The app scaled well on high DPI screens and the Fluent design is pretty spot on.

# Things that need work
- **Improvements to the documentation:** This has been part of the frustration when making the app. Some details aren't documented (multiple blur settings for the window, the Title control, Adding the FilePickerFilters, etc.) or poorly documented (File Picker). If it wasn't for the Telegram group, I wouldn't be able to bring over code I had (since the app is expanded version of an open source one I also made).
- **The VS extension needs work:** Another bit of frustration is that the VS extension felt a bit jank. Intellisense sometimes fails (it usually ignores x:Static bindings, which are required for localisation), adding resx files were a bit error prone (getting errors for these until I could get it to work) and a few missing quality of life features that are on WPF aren't on Avalonia.
- **The Direct2D library is a bit broken:** Debugging an AvaloniaUI that uses Direct2D works, but if I try to publish, it throws build errors. I'll need to look at it more, but it kind of sucks that I can't use it (I'm fine using Skia for now).

# Some things I'd like to see
- **Some features from WPF editor brought over:** I think having a Properties window would be nice to have.
- **Improvements to the documentation**: If there is one thing to look into, is adding some extra information in the documentation (such as showing how to add custom file filters). I'd be happy to contribute (although, my English a bit rough here and there).
- **Port over some of the controls from the WinUI Community Toolkit:** This one is low on my list of improvements, but I see some controls in said community toolkit that could be useful (I'd like to see a version of the simplified ribbon style menu (can't recall it's name)).

Overall, I'm pretty happy with using Avalonia with future projects. There are some rough edges, but I'd add this to my toolbox.

**Comment by thevortexcloud** [comment]
If you are willing to spend some money you can buy Resharper/Rider  which has it's own intellisense for Avalonia. It works fairly well. As for the documentation, I agree. That said you can often instead search for WPF examples instead of Avalonia specific ones. It generally yields better results and is just as relevant 6/10 times.

**Comment by peter0302** [comment]
Too many people love Avalonia for me to say that no one should use it, but I can confidently say if you're coming from WPF, unless you absolutely need cross-platform, either use WinUI3 or just stick with WPF.

In addition to all the things @FirehawkV21 discusses, which I completely agree with, there's one huge original and unforgiveable sin that I can't get past. They deviated from WPF so much, and in so many places, and the vast majority of time completely unnecessarily, that it is like learning an entirely new framework.

Take Styling. Using CSS-like selectors is definitely interesting. One might argue it's an improvement. But it's confusing A.F. until you get the hang of it, and extremely hard to debug. If you're trying to port anything from a different XAML flavor, you can forget about it.

Or something as simple as `DependencyProperty`. Did it really have to be called `AvaloniaProperty` instead? Would it really have killed the dream to make those classes and registration methods match WPF? I have hundreds of controls with thousands of DPs. Do I really want to be spending my time renaming them all for utterly no reason except someone liked the name better? 

But here's the final insult added to the injury -  there IS a flavor that's WPF compatible, but you have to pay for it. No thanks. I don't pay for development frameworks, sorry. Whoever their investors are who have the lame idea they can actually montize this somehow or that Microsoft is gonna acquire them, they can dream on. Bless their hearts.

So yeah, I gave this a real college try. I truly did. But I'm a hard pass.

**Reply by maxkatz6** [reply]
> But here's the final insult added to the injury - there IS a flavor that's WPF compatible, but you have to pay for it. No thanks

XPF is not a flavor of Avalonia. It's an original WPF, with rebuilt core windowing and rendering subsystems, for which you need to pay. It's specifically focused for companies that need to stick to WPF, while making it crossplatform. For new apps we usually recommend Avalonia (free).

Other than that, Avalonia is not a WPF. And it doesn't have to be WPF compatible. You could have the same complain about XF/MAUI and their own dialect of XAML - but these also doesn't have to have the same XAML.

**Reply by peter0302** [reply]
> XPF is not a flavor of Avalonia. It's an original WPF, with rebuilt core windowing and rendering subsystems, for which you need to pay

I'm confused. Are you saying XPF is just the Microsoft WPF source code pulled from .NET Core but with a new rendering engine? Something I did for fun over a weekend a few years back? Again, no thanks. I want a modern framework, otherwise I'd just use WPF. I don't care about cross-platform.

> Other than that, Avalonia is not a WPF. And it doesn't have to be WPF compatible. 

Avalonia didn't have to be 100% API/XAML compatible with WPF. You could've simply not gone out of your way to make it maddeningly difficult for someone who just wants to move an existing codebase to a modern framework without spending thousands of completely unnecessary hours of syntactical refactoring and control restyling.

>  You could have the same complain about XF/MAUI and their own dialect of XAML - but these also doesn't have to have the same XAML.

And I do, but there's at least a history there. Xamarin never billed itself as an alternative to WPF or a desktop app framework originally. Avalonia did. The "spiritual successor to WPF." Not so much.

Look, you guys have a blast and keep doing what you're doing because you have a devoted fan base. But again, IMO anyone coming from WPF or looking to modernize an existing commercial product should just keep moving and not look back.

**Reply by jmacato** [reply]
```
They deviated from WPF so much, and in so many places, and the vast majority of time completely unnecessarily, that it is like learning an entirely new framework.
```

Because we *are* a new framework ;) 

```
just the Microsoft WPF source code pulled from .NET Core but with a new rendering engine? Something I did for fun over a weekend a few years back?
```

The proof's in the pudding buddy ;)

If you think pushing everyone's buttons will make your point across then bless your heart indeed.

**Reply by rabbitism** [reply]
>  No thanks. I don't pay for development frameworks, sorry. 

OKAY. Thanks for letting me know.

**Reply by thevortexcloud** [reply]
> They deviated from WPF so much, and in so many places

This is like complaining about electric cars for not having an internal combustion engine which prevents you from using your existing collection of engine oils. You are dealing with a fundamentally different product that just happens to work _similarly_ to something you used to use; that can also be used to accomplish most of the same tasks.

If you want to use WPF, you can use WPF. If you want WPF that runs on top of Avalonia's low level code then there is XPF.

Now if somebody wants to make a compatibility shim (ie implements WPF's API surface ontop of Avalonia without using any actual WPF code) then they are more than welcome to make that too.

**Comment by peter0302** [comment]
It’s more like Avalonia is an electric car that needlessly relabeled and redesigned all the controls of a normal car into something like Mr. Garrison’s IT Machine from South Park.

**Reply by grokys** [reply]
I realise that you're just trolling at this point, but to give you some context: when I started Avalonia it was me who made many of the decisions you're complaining about.

The simple reason that many things are different is that I thought it made sense at the time, and I never actually expected the project to get as popular as it has. Hell I never expected to be working on the project in 2 years time (this was 2013/2014), and I most certainly never expected to have more than a handful of users.

If I could go back in time, I'd reverse many of those decisions, sure. Turns out though, people started building on Avalonia and now those arbitrary decisions are somewhat set in stone.

**Reply by grokys** [reply]
What's the moral of the story? "Don't do things for fun, it might become popular" I guess.

**Reply by grokys** [reply]
While I'm here I may as well give some reasoning from my vague recollection of 10 years ago to a few of the points you pointed out:

> Or something as simple as `DependencyProperty`. Did it really have to be called `AvaloniaProperty` instead? Would it really have killed the dream to make those classes and registration methods match WPF?

If I remember correctly I did originally call it `DependencyProperty` but a few times while I was playing about with it, VS autocompleted `DependencyProperty` and added a `using System.Windows;` statement, pulling in the WPF one when I wanted the Avalonia one (though it was called Perspex back at that point[^1]). I probably didn't think much about it, sorry about that. That decision is indeed something I'd reverse (though I maintain that giving the property fields types was a good idea, so there would be some changes necessary anyway).

> Take Styling. Using CSS-like selectors is definitely interesting. One might argue it's an improvement. But it's confusing A.F. until you get the hang of it, and extremely hard to debug. If you're trying to port anything from a different XAML flavor, you can forget about it.

Simply put, I didn't like the way WPF styles were so limiting and thought that CSS styles were much more flexible. We have somewhat rolled back that change in 11 by introducing `ControlTheme`s but we had to use a different naming to WPF due to their naming clashing. If I could go back in time I'd name the CSS-like styles something else.

[^1]: Another thing I'd do if I could go back in time was not use a trademarked name as the name of the project.

**Reply by peter0302** [reply]
A joke <> trolling. I'm trying to give you all feedback as to why I as the owner of a very large WPF project, which is commercially successful, who's been looking for a new framework for our next gen app for years, and gave yours an honest shakedown, ultimately found it to be unsuitable.

But I appreciate what you're saying. We're all beholden to technical debt. However rather than fork WPF as it sounds like you did for XPF, I would've preferred a fork of Avalonia but refactored with the WPF syntax whenever feasible. That I would even have considered paying for provided it was open source and I was only paying for support. But again I don't want WPF with a new native rendering engine. The managed portion of WPF is also a nightmare under the hood as I'm sure you know. That's what almost 20 years without a meaningful overhaul gets you and no doubt why MS abandoned it and re-wrote the whole thing in C++ for RT. But the black box was largely fine and IMO needed minimal changes. Indeed for RT, with all its faults, they made far, far fewer deviations from WPF syntax than you did.

Or if you really want to be superstars, fork the finally open source WinUI3 in C++ and port _that_ to other platforms like WASM or Mac. Holy cow would that be something to behold.

The electric car analogy was apt. I would love a snappy sleek beautiful electirc car. But not if the dashboard and pedals are unrecognizable.

There's no moral. You went from being a fun popular project to a commercial enterprise that's taken very seriously by other businesses. So people like me are gonna judge you by that standard, including with feedback you're not gonna love. It's not personal. As a fun open source project everything I've seen indicates Avalonia is great. It's just not for us.

**Reply by grokys** [reply]
Absolutely no offense taken, I agree with many of your points in fact.

> fork the finally open source WinUI3 in C++ and port that to other platforms like WASM or Mac.

I think this is pretty much exactly what the https://platform.uno/ folks are doing. Maybe check them out?
